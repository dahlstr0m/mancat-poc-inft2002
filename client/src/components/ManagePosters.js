// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import Button from './Button';
import Form from './Form';
import { Card } from './Card';
import { Row, Column, Alert } from './Widgets';
import projectService, { type Project } from '../services/project-service';
import posterService, { type Poster } from '../services/poster-service';

/**
 * Manage posters page
 */
export default class ManagePosters extends Component {
  poster: Poster = {
    // Default values for poster, updated on change in form
    posterId: 1,
    projectId: 1,
    posterDescription: '',
    posterUrl: '',
    thumbnailUrl: '',
  };
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        <Card title="Edit existing poster">
          <Form.Label> Choose project:</Form.Label>
          <Form.Select
            value={this.poster.projectId}
            onChange={(event) => (this.poster.projectId = parseInt(event.currentTarget.value))}
          >
            {this.projects.map((project) => (
              <option key={project.projectId} value={project.projectId}>
                {project.title}
              </option>
            ))}
          </Form.Select>
          <Card title="These posters exists for this project. Click to edit:">
            {this.posters
              .filter((poster) => poster.projectId == this.poster.projectId)
              .map((poster) => (
                <Row key={poster.posterId}>
                  <Column width={2}>
                    <img
                      src={poster.thumbnailUrl}
                      alt={'Can not display poster with id ' + poster.posterId}
                      width="105"
                      heigth="150"
                    />
                  </Column>
                  <Column>
                    <Card>
                      <Form.Label> Poster description:</Form.Label>
                      <Form.Input
                        type="text"
                        value={poster.posterDescription}
                        onChange={(event) => (poster.posterDescription = event.currentTarget.value)}
                      ></Form.Input>
                      <br />
                      <Form.Label> Url to Poster:</Form.Label>
                      <Form.Input
                        type="text"
                        value={poster.posterUrl}
                        onChange={(event) => (poster.posterUrl = event.currentTarget.value)}
                      ></Form.Input>
                      <br />
                      <Form.Label> Url to posters thumbnail:</Form.Label>
                      <Form.Input
                        type="text"
                        value={poster.thumbnailUrl}
                        onChange={(event) => (poster.thumbnailUrl = event.currentTarget.value)}
                      ></Form.Input>
                      <br />
                      <Button.Success
                        onClick={() =>
                          posterService
                            .updatePoster(poster)
                            .then(() => {
                              history.push('/admin/');
                              Alert.success('Poster successfully edited');
                            })
                            .catch((error: Error) =>
                              Alert.danger('Error storing edited poster: ' + error.message)
                            )
                        }
                      >
                        Save edit
                      </Button.Success>
                      <Button.Danger
                        onClick={() =>
                          posterService
                            .deletePoster(poster.posterId)
                            .then(() => {
                              history.push('/admin');
                              Alert.success('Poster successfully deleted');
                            })
                            .catch((error: Error) =>
                              Alert.danger('Error deleting poster: ' + error.message)
                            )
                        }
                      >
                        Delete poster
                      </Button.Danger>
                    </Card>
                  </Column>
                </Row>
              ))}
          </Card>
          <br />
        </Card>
      </>
    );
  }
  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting employers: ' + error.message));
    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error fetching connected posters: ' + error.message));
  }
}
