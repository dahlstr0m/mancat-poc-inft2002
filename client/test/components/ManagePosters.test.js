// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import ManagePosters from '../../src/components/ManagePosters';

// Mocks
jest.mock('../../src/services/project-service');
jest.mock('../../src/services/poster-service');

// Tests
describe('ManagePosters component tests', () => {
  test('ManagePosters draws correctly', (done) => {
    const wrapper = shallow(<ManagePosters />);

    // Wait for events to complete
    setTimeout(() => {
      setTimeout(() => {
        expect(wrapper).toMatchSnapshot();

        done();
      });
    });
  });
});
