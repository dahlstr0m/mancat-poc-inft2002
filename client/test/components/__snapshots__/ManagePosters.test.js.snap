// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagePosters component tests ManagePosters draws correctly 1`] = `
<Fragment>
  <Card
    title="Edit existing poster"
  >
    <FormLabel>
       Choose project:
    </FormLabel>
    <FormSelect
      onChange={[Function]}
      value={1}
    >
      <option
        key="1"
        value={1}
      >
        Project 1
      </option>
      <option
        key="2"
        value={2}
      >
        Project 2
      </option>
      <option
        key="3"
        value={3}
      >
        Project 3
      </option>
    </FormSelect>
    <Card
      title="These posters exists for this project. Click to edit:"
    >
      <Row
        key="1"
      >
        <Column
          width={2}
        >
          <img
            alt="Can not display poster with id 1"
            heigth="150"
            src="Poster 1 thumb url"
            width="105"
          />
        </Column>
        <Column>
          <Card>
            <FormLabel>
               Poster description:
            </FormLabel>
            <FormInput
              onChange={[Function]}
              type="text"
              value="Poster 1 description"
            />
            <br />
            <FormLabel>
               Url to Poster:
            </FormLabel>
            <FormInput
              onChange={[Function]}
              type="text"
              value="Poster 1 url"
            />
            <br />
            <FormLabel>
               Url to posters thumbnail:
            </FormLabel>
            <FormInput
              onChange={[Function]}
              type="text"
              value="Poster 1 thumb url"
            />
            <br />
            <ButtonSuccess
              onClick={[Function]}
            >
              Save edit
            </ButtonSuccess>
            <ButtonDanger
              onClick={[Function]}
            >
              Delete poster
            </ButtonDanger>
          </Card>
        </Column>
      </Row>
    </Card>
    <br />
  </Card>
</Fragment>
`;
